Case importer field mapping improvements

dannyroberts opened the pull request:
- Adds a notion of special fields, statically defined (like `owner_name`, etc.)
- Shows description of special fields in the UI if you use them
- Admonishes you if you use a non "discoverable" property (currently just `parent_ref`)
- Adds a "Did you mean?" feature for the case field, which suggests any case property or discoverable special field
- Switches interactive part of page to knockout
- Gets rid of jumpiness in case field input box
- Small, but make 'name' show in the menu and be recognized as a known property
- Autofill by default (without you having to press autofill); change name to "Set to default".

<img width="974" alt="screen shot 2016-12-15 at 10 36 50 pm" src="https://cloud.githubusercontent.com/assets/137212/21250929/0960d1de-c317-11e6-9f27-da33a1a615aa.png">

EDIT Dec 19, 2016: The final version of this feature also

- makes 'external_id' a special field in the UI (so it comes up with a description)
- makes the "did you mean" functionality case-insensitive (so it'll still suggest "owner_name" for "OWNER NAME")

dannyroberts committed:
document case importer special fields in code

and switch to using that, without changing current functionality

dannyroberts committed:
treat 'name' as a known case field in case importer

dannyroberts committed:
refactor interaction-heavy portion of excel_fields.html

to use knockout. Currently some features broken, but main interactions working

dannyroberts committed:
ko excel_fields: remove jquery checkbox control to ko

dannyroberts committed:
ko excel_fields: move removeRow to ko

dannyroberts committed:
ko excel_fields: move add mapping to ko

dannyroberts committed:
ko excel_field: move placement and styling of special field description

dannyroberts committed:
fix typo

dannyroberts committed:
mark parent_ref case field non-discoverable

dannyroberts committed:
excel_field: improve styling for special fields

dannyroberts committed:
excel fields: autofill by default and move to ko

dannyroberts committed:
fix jquery click

dannyroberts committed:
excel fields: add "Did you mean ...?"

dannyroberts committed:
excel fields: fix jumpiness in case field inputs

dannyroberts committed:
fix suggestion order

dannyroberts committed:
fix js lint error

dannyroberts committed:
systematize sanitizing case field

dannyroberts committed:
remove debugging comments

dimagimon commented on corehq/apps/case_importer/static/case_importer/js/excel_fields.js:
```diff
@@ -0,0 +1,89 @@
+                    return [];
+                }
+                var suggestions = _(self.caseFieldSuggestions).map(function (suggestion) {
+                    return {distance: Levenshtein.get(field, suggestion), field: suggestion};
```
'Levenshtein' is not defined. (no-undef)

benrudolph replied:
nit: can use `window.Levenshtein`

dimagimon replied:
'Levenshtein' is not defined. (no-undef)

dannyroberts commented:
Ready for review, but I'm going to test this a bit more in the morning with fresh brain.

benrudolph commented on corehq/apps/case_importer/static/case_importer/js/excel_fields.js:
```diff
@@ -0,0 +1,89 @@
+                var suggestions = _(self.caseFieldSuggestions).map(function (suggestion) {
+                    return {distance: Levenshtein.get(field, suggestion), field: suggestion};
+                }).filter(function (suggestion) {
+                    return suggestion.distance < 4;
```
if the distance is 4 maybe we should do non case sensitive levenshtein? otherwise `OWNER_names` will be distance 6 away from `owner_name`. not sure how to best handle since i think the backend depends on the case sensitivity of the field. pretty minor point

dannyroberts replied:
That's a fair point, and I think pretty easy to implement without making the whole field case insensitive. Just change the distance calculation to be `Levenshtein.get(field.toLowerCase(), suggestion.toLowerCase())` so that the distance between what they type and a suggestion only takes into account how far apart their case-insensitive forms are. Honestly don't know how often people use all-caps, but it's easy and harmless to accommodate.

benrudolph commented on corehq/apps/case_importer/util.py:
```diff
@@ -424,6 +424,9 @@ def get_from_location(name):
 def get_case_properties_for_case_type(domain, case_type):
+    # todo: seems like poor boundaries for this function care about the backend
+    # todo: get_case_properties just always return the right answer,
+    # todo: possibly by moving this there.
```
were you planning on addressing this?

benrudolph commented on corehq/apps/case_importer/views.py:
```diff
@@ -195,24 +196,14 @@ def excel_fields(request, domain):
-    try:
-        case_fields.remove('type')
-    except:
-        pass
```
😖

dannyroberts committed:
Merge branch 'master' into case-importer-did-you-mean

benrudolph reviewed (approved):
❤️ this! good cleanup, only minor comments. it it's easy to test, i might add a few in there

dannyroberts commented:
I committed the latest dimagi-utils to master (https://github.com/dimagi/commcare-hq/commit/3d6336fdaf8a8b8466c552074d7be6607b7f5384) and merged it in here.

dannyroberts committed:
excel fields: make external_id a speical property

- fix edge cases around that
- do not exclude search_column, since that refers to the excel colum and not the case field

dimagimon commented on corehq/apps/case_importer/tests/test_suggested_fields.py:
```diff
@@ -0,0 +1,60 @@
+import mock
+from corehq.apps.case_importer.suggested_fields import get_suggested_case_fields, \
+    FieldSpec
+from corehq.util.test_utils import DocTestMixin, generate_cases
```
F401 'generate_cases' imported but unused

dannyroberts committed:
use window.* for explicitly referring to a global

dannyroberts committed:
make case importer did-you-know case-insensitive

dimagimon commented on corehq/apps/case_importer/static/case_importer/js/excel_fields.js:
```diff
@@ -0,0 +1,93 @@
+                    return {
+                        // make distance case-insensitive
+                        distance: window.Levenshtein.get(field.toLowerCase(), suggestion.toLowerCase()),
+                        field: suggestion
```
Missing trailing comma. (comma-dangle)

dannyroberts committed:
linter feedback

dannyroberts commented:
@benrudolph anything in particular you thought I should test? I added a test as I was making another change just now, but haven't added any javascript test, etc.

dannyroberts committed:
test excel_fields.js:sanitizeCaseField

dimagimon commented:
There are 8 errors:

* corehq/apps/case_importer/static/case_importer/spec/excel_fields_spec.js, line 1 - 'describe' is not defined. (no-undef)
* corehq/apps/case_importer/static/case_importer/spec/excel_fields_spec.js, line 3 - 'describe' is not defined. (no-undef)
* corehq/apps/case_importer/static/case_importer/spec/excel_fields_spec.js, line 4 - 'it' is not defined. (no-undef)
* corehq/apps/case_importer/static/case_importer/spec/excel_fields_spec.js, line 7 - 'it' is not defined. (no-undef)
* corehq/apps/case_importer/static/case_importer/spec/excel_fields_spec.js, line 10 - 'it' is not defined. (no-undef)
* corehq/apps/case_importer/static/case_importer/spec/excel_fields_spec.js, line 13 - 'it' is not defined. (no-undef)
* corehq/apps/case_importer/static/case_importer/spec/excel_fields_spec.js, line 16 - 'it' is not defined. (no-undef)
* corehq/apps/case_importer/static/case_importer/spec/excel_fields_spec.js, line 19 - Missing semicolon. (semi)

dannyroberts committed:
fix typo search_field => search_column

dannyroberts committed:
use .ko-template to fix page load flashing

esoergel commented on corehq/apps/case_importer/suggested_fields.py:
```diff
@@ -0,0 +1,88 @@
+    """
+    take a list of FieldSpec objects and return a sorted list where field is unique
+    and fields in exclude_fields are removed, and where the first mention of a
+    field in field_specs will win out over any repeats.
```
This behavior feels a little cryptic, couldn't you just use `dict.update` and apply `special_field_specs` last?

dannyroberts replied:
Yeah it felt like it was going to be that simple, but we're doing a few things at the same time. Maybe there was a way to do this more clearly, but I went with normalizing the inputs in the outer function (so that the special fields and the dynamic fields were both lists of FieldSpecs, and exclude was a set) and then doing the merge/unique-by-field/first-wins/exclude/sort thing in an inner function with a doc string that said what all was going on. Trying to do them together in a couple list or dict comprehensions the way I started with just turned out to be more confusing than it needed to be.

If you were just saying I should rewrite this function to take two lists instead of one, and then do

```
combined_field_specs = {}
combined_field_specs.update({field_spec.field for field_spec in second_list})
combined_field_specs.update({field_spec.field for field_spec in first_list})
combined_field_specs = {key: value for key, value in combined_field_specs.items() if key not in exclude_fields}
return sorted(combined_field_specs.values(), key=lambda field_spec: field_spec.field)
```

I could do that, but honestly I'm not sure it's that much better one way or the other.

esoergel replied:
Sounds like you've already thought through that approach, so I'm happy with what you've got here.

esoergel commented on corehq/apps/case_importer/suggested_fields.py:
```diff
@@ -0,0 +1,88 @@
+            field='close',
+            description=_("This field will be used to close cases. "
+                          "Any case with 'yes' in this column will be closed.")),
+    ]
```
I really like having these all spelled out like this!

dannyroberts replied:
Yeah, I started this whole thing by just writing out these specs for myself, before even having the feature for showing them to the user. It just felt kind of crazy that there were like 5 different places where special magic words were used in `case_importer` and no place that defined them as special (or bothered to mention this to the user!).

esoergel commented on bower.json:
```diff
@@ -63,7 +63,8 @@
     "At.js": "~1.5.2",
-    "clipboard": "1.5.15"
+    "clipboard": "1.5.15",
+    "fast-levenshtein": "^2.0.5"
```
Nice, I imagine there are a number of places we could use this.

esoergel commented on corehq/apps/case_importer/static/case_importer/spec/excel_fields_spec.js:
```diff
@@ -0,0 +1,22 @@
+            excelfields.sanitizeCaseField('my   property  name', 'my_property_name');
+        });
+    });
+});
```
:heart:

dannyroberts replied:
:)

czue commented:
this is really awesome!

dannyroberts commented:
Ok, this is good to merge by me now, tested on staging
